python/adelphi/adelphi/anonymize.py

```diff
@@ -38,59 +33,70 @@ def get_name(original_name, prefix):
 def anonymize_keyspace(keyspace):
     keyspace.name = get_name(keyspace.name, KEYSPACE_PREFIX)
 
+    # anonymize udts at once because they can reference
+    # one another
+    udts = keyspace.user_types.values()
+    anonymize_udts(udts)
+
     for table in keyspace.tables.values():
         anonymize_table(table)
 
-    for udt in keyspace.user_types.values():
-        anonymize_udt(udt)
+    # remove functions, aggregates and views for now
+    keyspace.functions = {}
+    keyspace.aggregates = {}
+    keyspace.views = {}
 
+def anonymize_type(original_type):
+    name = original_type
+    for udt in name_map[TYPE_PREFIX]:
+        name = name.replace(udt, name_map[TYPE_PREFIX][udt])
```

Comment:
What this does is replace all occurrences of an UDT name by its anonymized name (it applies to column types and fields types/nested UDTs).

Reply:
There's an explicit assumption here that name_map[TYPE_PREFIX] exists at this point, something that get_name does _not_ assume.  That assumption holds, but it only holds because of something that happens outside of this function... indeed, something that happens at a level _above_ this function.  That happens when we loop over all UDTs and populate their names first before this function is called.  All of that happens in anonymize_udts.

This sort of unenforced and undocumented dependency is almost always a recipe for silent breakage on future changes.  It would be nice if at a minimum we could make that dependency more explicit... and ideally remove it all together.

Reply:
Notice this is not only used for UDTs, it's also used by `anonymize_column`, so this is here mostly to avoid code duplication.
I think the only way to make the dependency more explicit would be to lump everything together in `anonymize_keyspace` and accept some code duplication.

> That happens when we loop over all UDTs and populate their names first before this function is called.
> This sort of unenforced and undocumented dependency is almost always a recipe for silent breakage on future changes. 

If someone ever changes that code path to anonymize columns or fields before UDTs, the unit tests included in this PR will immediately fail.

Reply:
Understood... I'd just prefer for the code to be more explicit about what it's doing there and how it handles those cases.  But given time constraints I don't feel like this is essential enough to hold up getting the larger set of improvements provided by this PR in so let's just skip it for now.